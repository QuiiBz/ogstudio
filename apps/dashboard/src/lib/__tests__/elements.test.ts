import { describe, it, expect } from "vitest";
import {
  createElementId,
  createElement,
  createElementStyle,
  createImgElementStyle,
} from "../elements";

describe("createElementId", () => {
  it("should generate a random id", () => {
    const id = createElementId();
    expect(id).to.be.a("string");
    expect(id.length).toBe(9);

    const id2 = createElementId();
    expect(id2).not.toEqual(id);
  });
});

describe("createElement", () => {
  it("should fill id, x and y fields based on width and height", () => {
    const element = createElement({
      tag: "div",
      width: 100,
      height: 200,
    });

    expect(element.id).to.be.a("string");
    expect(element.id.length).toBe(9);

    // @ts-expect-error not sure what's going on here
    delete element.id;

    expect(element).toMatchInlineSnapshot(`
      {
        "height": 200,
        "tag": "div",
        "width": 100,
        "x": 550,
        "y": 215,
      }
    `);
  });
});

describe("createElementStyle", () => {
  it("should generate base style", () => {
    const style = createElementStyle({
      tag: "div",
      id: createElementId(),
      name: "Box",
      x: 10,
      y: 20,
      width: 100,
      height: 200,
      visible: true,
      rotate: 0,
      blur: 0,
      backgroundColor: "#ffffff",
    });

    expect(style).toMatchInlineSnapshot(`
      {
        "background": "rgba(255, 255, 255, 1)",
        "display": "flex",
        "height": "200px",
        "left": "10px",
        "position": "absolute",
        "top": "20px",
        "width": "100px",
      }
    `);
  });

  it("should generate base style with rotation", () => {
    const style = createElementStyle({
      tag: "div",
      id: createElementId(),
      name: "Box",
      x: 10,
      y: 20,
      width: 100,
      height: 200,
      visible: true,
      rotate: 120,
      blur: 0,
      backgroundColor: "#ffffff",
    });

    expect(style).toMatchInlineSnapshot(`
      {
        "background": "rgba(255, 255, 255, 1)",
        "display": "flex",
        "height": "200px",
        "left": "10px",
        "position": "absolute",
        "top": "20px",
        "transform": "rotate(120deg)",
        "width": "100px",
      }
    `);
  });

  it("should generate base style with blur", () => {
    const style = createElementStyle({
      tag: "div",
      id: createElementId(),
      name: "Box",
      x: 10,
      y: 20,
      width: 100,
      height: 200,
      visible: true,
      rotate: 0,
      blur: 10,
      backgroundColor: "#ffffff",
    });

    expect(style).toMatchInlineSnapshot(`
      {
        "background": "rgba(255, 255, 255, 1)",
        "display": "flex",
        "filter": "blur(10px)",
        "height": "200px",
        "left": "10px",
        "position": "absolute",
        "top": "20px",
        "width": "100px",
      }
    `);
  });

  describe("border", () => {
    it("should generate base style with border outside", () => {
      const style = createElementStyle({
        tag: "div",
        id: createElementId(),
        name: "Box",
        x: 10,
        y: 20,
        width: 100,
        height: 200,
        visible: true,
        rotate: 0,
        blur: 0,
        border: {
          color: "blue",
          width: 2,
          style: "outside",
        },
        backgroundColor: "#ffffff",
      });

      expect(style).toMatchInlineSnapshot(`
        {
          "background": "rgba(255, 255, 255, 1)",
          "boxShadow": "0 0 0 2px blue",
          "display": "flex",
          "height": "200px",
          "left": "10px",
          "position": "absolute",
          "top": "20px",
          "width": "100px",
        }
      `);
    });

    it("should generate base style with border inside", () => {
      const style = createElementStyle({
        tag: "div",
        id: createElementId(),
        name: "Box",
        x: 10,
        y: 20,
        width: 100,
        height: 200,
        visible: true,
        rotate: 0,
        blur: 0,
        border: {
          color: "green",
          width: 4,
          style: "inside",
        },
        backgroundColor: "#ffffff",
      });

      expect(style).toMatchInlineSnapshot(`
        {
          "background": "rgba(255, 255, 255, 1)",
          "boxShadow": "0 0 0 4px inset green",
          "display": "flex",
          "height": "200px",
          "left": "10px",
          "position": "absolute",
          "top": "20px",
          "width": "100px",
        }
      `);
    });
  });

  it("should generate base style with shadow", () => {
    const style = createElementStyle({
      tag: "div",
      id: createElementId(),
      name: "Box",
      x: 10,
      y: 20,
      width: 100,
      height: 200,
      visible: true,
      rotate: 0,
      blur: 0,
      shadow: {
        color: "blue",
        width: 2,
        blur: 4,
        x: 2,
        y: 2,
      },
      backgroundColor: "#ffffff",
    });

    expect(style).toMatchInlineSnapshot(`
      {
        "background": "rgba(255, 255, 255, 1)",
        "boxShadow": "2px 2px 4px 2px blue",
        "display": "flex",
        "height": "200px",
        "left": "10px",
        "position": "absolute",
        "top": "20px",
        "width": "100px",
      }
    `);
  });

  it("should generate base style with border and shadodw", () => {
    const style = createElementStyle({
      tag: "div",
      id: createElementId(),
      name: "Box",
      x: 10,
      y: 20,
      width: 100,
      height: 200,
      visible: true,
      rotate: 0,
      blur: 0,
      border: {
        color: "green",
        width: 4,
        style: "inside",
      },
      shadow: {
        color: "blue",
        width: 2,
        blur: 4,
        x: 2,
        y: 2,
      },
      backgroundColor: "#ffffff",
    });

    expect(style).toMatchInlineSnapshot(`
      {
        "background": "rgba(255, 255, 255, 1)",
        "boxShadow": "0 0 0 4px inset green, 2px 2px 4px 2px blue",
        "display": "flex",
        "height": "200px",
        "left": "10px",
        "position": "absolute",
        "top": "20px",
        "width": "100px",
      }
    `);
  });

  describe("p & span", () => {
    it("should generate style", () => {
      const style = createElementStyle({
        tag: "p",
        id: createElementId(),
        name: "Box",
        x: 10,
        y: 20,
        width: 100,
        height: 200,
        visible: true,
        rotate: 0,
        blur: 0,
        content: "Hello",
        color: "#ffffff",
        fontFamily: "Inter",
        fontSize: 12,
        fontWeight: 400,
        lineHeight: 1,
        letterSpacing: 0,
        align: "left",
      });

      expect(style).toMatchInlineSnapshot(`
        {
          "color": "rgba(255, 255, 255, 1)",
          "display": "flex",
          "fontFamily": "Inter",
          "fontSize": "12px",
          "fontWeight": 400,
          "height": "200px",
          "justifyContent": "flex-start",
          "left": "10px",
          "letterSpacing": "0px",
          "lineHeight": 1,
          "marginBottom": 0,
          "marginTop": 0,
          "position": "absolute",
          "top": "20px",
          "width": "100px",
        }
      `);
    });

    it("should generate style with shadow", () => {
      const style = createElementStyle({
        tag: "p",
        id: createElementId(),
        name: "Box",
        x: 10,
        y: 20,
        width: 100,
        height: 200,
        visible: true,
        rotate: 0,
        blur: 0,
        shadow: {
          color: "blue",
          width: 2,
          blur: 4,
          x: 2,
          y: 2,
        },
        content: "Hello",
        color: "#ffffff",
        fontFamily: "Inter",
        fontSize: 12,
        fontWeight: 400,
        lineHeight: 1,
        letterSpacing: 0,
        align: "left",
      });

      expect(style).toMatchInlineSnapshot(`
        {
          "color": "rgba(255, 255, 255, 1)",
          "display": "flex",
          "fontFamily": "Inter",
          "fontSize": "12px",
          "fontWeight": 400,
          "height": "200px",
          "justifyContent": "flex-start",
          "left": "10px",
          "letterSpacing": "0px",
          "lineHeight": 1,
          "marginBottom": 0,
          "marginTop": 0,
          "position": "absolute",
          "textShadow": "2px 2px 4px blue",
          "top": "20px",
          "width": "100px",
        }
      `);
    });

    it("should generate style with align center", () => {
      const style = createElementStyle({
        tag: "p",
        id: createElementId(),
        name: "Box",
        x: 10,
        y: 20,
        width: 100,
        height: 200,
        visible: true,
        rotate: 0,
        blur: 0,
        content: "Hello",
        color: "#ffffff",
        fontFamily: "Inter",
        fontSize: 12,
        fontWeight: 400,
        lineHeight: 1,
        letterSpacing: 0,
        align: "center",
      });

      expect(style).toMatchInlineSnapshot(`
        {
          "color": "rgba(255, 255, 255, 1)",
          "display": "flex",
          "fontFamily": "Inter",
          "fontSize": "12px",
          "fontWeight": 400,
          "height": "200px",
          "justifyContent": "center",
          "left": "10px",
          "letterSpacing": "0px",
          "lineHeight": 1,
          "marginBottom": 0,
          "marginTop": 0,
          "position": "absolute",
          "top": "20px",
          "width": "100px",
        }
      `);
    });

    it("should generate style with align right", () => {
      const style = createElementStyle({
        tag: "p",
        id: createElementId(),
        name: "Box",
        x: 10,
        y: 20,
        width: 100,
        height: 200,
        visible: true,
        rotate: 0,
        blur: 0,
        content: "Hello",
        color: "#ffffff",
        fontFamily: "Inter",
        fontSize: 12,
        fontWeight: 400,
        lineHeight: 1,
        letterSpacing: 0,
        align: "right",
      });

      expect(style).toMatchInlineSnapshot(`
        {
          "color": "rgba(255, 255, 255, 1)",
          "display": "flex",
          "fontFamily": "Inter",
          "fontSize": "12px",
          "fontWeight": 400,
          "height": "200px",
          "justifyContent": "flex-end",
          "left": "10px",
          "letterSpacing": "0px",
          "lineHeight": 1,
          "marginBottom": 0,
          "marginTop": 0,
          "position": "absolute",
          "top": "20px",
          "width": "100px",
        }
      `);
    });
  });

  describe("div", () => {
    it("should generate style with radius", () => {
      const style = createElementStyle({
        tag: "div",
        id: createElementId(),
        name: "Box",
        x: 10,
        y: 20,
        width: 100,
        height: 200,
        visible: true,
        rotate: 0,
        blur: 0,
        radius: 4,
        backgroundColor: "#ffffff",
      });

      expect(style).toMatchInlineSnapshot(`
        {
          "background": "rgba(255, 255, 255, 1)",
          "borderRadius": "4px",
          "display": "flex",
          "height": "200px",
          "left": "10px",
          "position": "absolute",
          "top": "20px",
          "width": "100px",
        }
      `);
    });

    describe("gradient", () => {
      it("should generate style with linear gradient", () => {
        const style = createElementStyle({
          tag: "div",
          id: createElementId(),
          name: "Box",
          x: 10,
          y: 20,
          width: 100,
          height: 200,
          visible: true,
          rotate: 0,
          blur: 0,
          backgroundColor: "#ffffff",
          gradient: {
            start: "#ff0000",
            end: "#00ff00",
            angle: 45,
            type: "linear",
          },
        });

        expect(style).toMatchInlineSnapshot(`
          {
            "background": "linear-gradient(45deg, #ff0000, #00ff00)",
            "display": "flex",
            "height": "200px",
            "left": "10px",
            "position": "absolute",
            "top": "20px",
            "width": "100px",
          }
        `);
      });

      it("should generate style with radial gradient", () => {
        const style = createElementStyle({
          tag: "div",
          id: createElementId(),
          name: "Box",
          x: 10,
          y: 20,
          width: 100,
          height: 200,
          visible: true,
          rotate: 0,
          blur: 0,
          backgroundColor: "#ffffff",
          gradient: {
            start: "#ff0000",
            end: "#00ff00",
            angle: 0,
            type: "radial",
          },
        });

        expect(style).toMatchInlineSnapshot(`
          {
            "background": "radial-gradient(#ff0000, #00ff00)",
            "display": "flex",
            "height": "200px",
            "left": "10px",
            "position": "absolute",
            "top": "20px",
            "width": "100px",
          }
        `);
      });
    });
  });
});

describe("createImgElementStyle", () => {
  it("should generate style with image and cover", () => {
    const style = createImgElementStyle({
      tag: "div",
      id: createElementId(),
      name: "Box",
      x: 10,
      y: 20,
      width: 100,
      height: 200,
      visible: true,
      rotate: 0,
      blur: 0,
      backgroundColor: "#ffffff",
      backgroundImage: "https://via.placeholder.com/150",
      backgroundSize: "cover",
    });

    expect(style).toMatchInlineSnapshot(`
      {
        "objectFit": "cover",
      }
    `);
  });

  it("should generate style with image and contain", () => {
    const style = createImgElementStyle({
      tag: "div",
      id: createElementId(),
      name: "Box",
      x: 10,
      y: 20,
      width: 100,
      height: 200,
      visible: true,
      rotate: 0,
      blur: 0,
      backgroundColor: "#ffffff",
      backgroundImage: "https://via.placeholder.com/150",
      backgroundSize: "contain",
    });

    expect(style).toMatchInlineSnapshot(`
      {
        "objectFit": "contain",
      }
    `);
  });

  it("should generate style with radius", () => {
    const style = createImgElementStyle({
      tag: "div",
      id: createElementId(),
      name: "Box",
      x: 10,
      y: 20,
      width: 100,
      height: 200,
      visible: true,
      rotate: 0,
      blur: 0,
      radius: 10,
      backgroundColor: "#ffffff",
      backgroundImage: "https://via.placeholder.com/150",
      backgroundSize: "cover",
    });

    expect(style).toMatchInlineSnapshot(`
      {
        "borderRadius": "10px",
        "objectFit": "cover",
      }
    `);
  });
});
