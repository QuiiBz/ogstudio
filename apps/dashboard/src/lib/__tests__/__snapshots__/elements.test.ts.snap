// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getImageElementSrc > should return data svg if the image is an SVG 1`] = `"data:image/svg+xml,%3Csvg%20width%3D%221728%22%20height%3D%221118%22%20viewBox%3D%220%200%201728%201118%22%20fill%3D%22none%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%20%3Cg%20clip-path%3D%22url(%23clip0_134_2)%22%3E%20%3Crect%20width%3D%221728%22%20height%3D%221118%22%20fill%3D%22white%22%2F%3E%20%3Cline%20x1%3D%22-4.37114e-08%22%20y1%3D%221117.5%22%20x2%3D%221728%22%20y2%3D%221117.5%22%20stroke%3D%22%23D1D1D1%22%2F%3E%20%3Cline%20x1%3D%22-4.37114e-08%22%20y1%3D%22279.5%22%20x2%3D%221728%22%20y2%3D%22279.5%22%20stroke%3D%22%23D1D1D1%22%2F%3E%20%3Cline%20x1%3D%22863.5%22%20y1%3D%221118%22%20x2%3D%22863.5%22%20y2%3D%22-2.18556e-08%22%20stroke%3D%22%23D1D1D1%22%2F%3E%20%3Cline%20x1%3D%221295.5%22%20y1%3D%221118%22%20x2%3D%221295.5%22%20y2%3D%22-2.18556e-08%22%20stroke%3D%22%23D1D1D1%22%2F%3E%20%3Cline%20x1%3D%22431.5%22%20y1%3D%221118%22%20x2%3D%22431.5%22%20y2%3D%22-2.18556e-08%22%20stroke%3D%22%23D1D1D1%22%2F%3E%20%3Cline%20x1%3D%22-4.37114e-08%22%20y1%3D%22838.5%22%20x2%3D%221728%22%20y2%3D%22838.5%22%20stroke%3D%22%23D1D1D1%22%2F%3E%20%3Cline%20x1%3D%22-4.37114e-08%22%20y1%3D%22558.5%22%20x2%3D%221728%22%20y2%3D%22558.5%22%20stroke%3D%22%23D1D1D1%22%2F%3E%20%3Cg%20opacity%3D%220.63%22%3E%20%3Cg%20filter%3D%22url(%23filter0_f_134_2)%22%3E%20%3Ccircle%20cx%3D%221560%22%20cy%3D%22-69%22%20r%3D%22253%22%20fill%3D%22%23F4E347%22%20fill-opacity%3D%220.4%22%2F%3E%20%3C%2Fg%3E%20%3Cg%20filter%3D%22url(%23filter1_f_134_2)%22%3E%20%3Ccircle%20cy%3D%221181%22%20r%3D%22253%22%20fill%3D%22%2317275D%22%20fill-opacity%3D%220.4%22%2F%3E%20%3C%2Fg%3E%20%3Cg%20filter%3D%22url(%23filter2_f_134_2)%22%3E%20%3Ccircle%20cx%3D%221794%22%20cy%3D%221258%22%20r%3D%22498%22%20fill%3D%22%2347EAF4%22%20fill-opacity%3D%220.4%22%2F%3E%20%3C%2Fg%3E%20%3Cg%20filter%3D%22url(%23filter3_f_134_2)%22%3E%20%3Ccircle%20cx%3D%22-52.5%22%20cy%3D%22-143.5%22%20r%3D%22396.5%22%20fill%3D%22%236AF447%22%20fill-opacity%3D%220.4%22%2F%3E%20%3C%2Fg%3E%20%3C%2Fg%3E%20%3C%2Fg%3E%20%3Cdefs%3E%20%3Cfilter%20id%3D%22filter0_f_134_2%22%20x%3D%221107%22%20y%3D%22-522%22%20width%3D%22906%22%20height%3D%22906%22%20filterUnits%3D%22userSpaceOnUse%22%20color-interpolation-filters%3D%22sRGB%22%3E%20%3CfeFlood%20flood-opacity%3D%220%22%20result%3D%22BackgroundImageFix%22%2F%3E%20%3CfeBlend%20mode%3D%22normal%22%20in%3D%22SourceGraphic%22%20in2%3D%22BackgroundImageFix%22%20result%3D%22shape%22%2F%3E%20%3CfeGaussianBlur%20stdDeviation%3D%22100%22%20result%3D%22effect1_foregroundBlur_134_2%22%2F%3E%20%3C%2Ffilter%3E%20%3Cfilter%20id%3D%22filter1_f_134_2%22%20x%3D%22-453%22%20y%3D%22728%22%20width%3D%22906%22%20height%3D%22906%22%20filterUnits%3D%22userSpaceOnUse%22%20color-interpolation-filters%3D%22sRGB%22%3E%20%3CfeFlood%20flood-opacity%3D%220%22%20result%3D%22BackgroundImageFix%22%2F%3E%20%3CfeBlend%20mode%3D%22normal%22%20in%3D%22SourceGraphic%22%20in2%3D%22BackgroundImageFix%22%20result%3D%22shape%22%2F%3E%20%3CfeGaussianBlur%20stdDeviation%3D%22100%22%20result%3D%22effect1_foregroundBlur_134_2%22%2F%3E%20%3C%2Ffilter%3E%20%3Cfilter%20id%3D%22filter2_f_134_2%22%20x%3D%221096%22%20y%3D%22560%22%20width%3D%221396%22%20height%3D%221396%22%20filterUnits%3D%22userSpaceOnUse%22%20color-interpolation-filters%3D%22sRGB%22%3E%20%3CfeFlood%20flood-opacity%3D%220%22%20result%3D%22BackgroundImageFix%22%2F%3E%20%3CfeBlend%20mode%3D%22normal%22%20in%3D%22SourceGraphic%22%20in2%3D%22BackgroundImageFix%22%20result%3D%22shape%22%2F%3E%20%3CfeGaussianBlur%20stdDeviation%3D%22100%22%20result%3D%22effect1_foregroundBlur_134_2%22%2F%3E%20%3C%2Ffilter%3E%20%3Cfilter%20id%3D%22filter3_f_134_2%22%20x%3D%22-649%22%20y%3D%22-740%22%20width%3D%221193%22%20height%3D%221193%22%20filterUnits%3D%22userSpaceOnUse%22%20color-interpolation-filters%3D%22sRGB%22%3E%20%3CfeFlood%20flood-opacity%3D%220%22%20result%3D%22BackgroundImageFix%22%2F%3E%20%3CfeBlend%20mode%3D%22normal%22%20in%3D%22SourceGraphic%22%20in2%3D%22BackgroundImageFix%22%20result%3D%22shape%22%2F%3E%20%3CfeGaussianBlur%20stdDeviation%3D%22100%22%20result%3D%22effect1_foregroundBlur_134_2%22%2F%3E%20%3C%2Ffilter%3E%20%3CclipPath%20id%3D%22clip0_134_2%22%3E%20%3Crect%20width%3D%221728%22%20height%3D%221118%22%20fill%3D%22white%22%2F%3E%20%3C%2FclipPath%3E%20%3C%2Fdefs%3E%20%3C%2Fsvg%3E"`;
